backend/services/user_groups.go

```diff
@@ -0,0 +1,441 @@
+	Slug      string
+	UserSlugs []string
+}
+
+type ModifyUserGroupInput struct {
+	Name          string
+	Slug          string
+	UsersToAdd    []string
+	UsersToRemove []string
+}
+
+type ListUserGroupsForAdminInput struct {
+	UserGroupFilter
+	Pagination
+	IncludeDeleted bool
+}
+
+type ListUserGroupsForOperationInput struct {
+	Pagination
+	UserGroupFilter
+	OperationSlug string
+}
+
+type userGroupAndRole struct {
+	models.UserGroup
+	Role policy.OperationRole `db:"role"`
+}
+
+type ListUserGroupsInput struct {
+	Query          string
+	IncludeDeleted bool
+	OperationSlug  string
+}
+
+func (cugi ModifyUserGroupInput) validateUserGroupInput() error {
+	if cugi.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	return nil
+}
+
+func (cugi CreateUserGroupInput) validateUserGroupInput() error {
+	if cugi.Slug == "" {
+		return backend.MissingValueErr("Slug")
+	}
+	if cugi.Name == "" {
+		return backend.MissingValueErr("Name")
+	}
+	return nil
+}
+
+func AddUsersToGroup(db database.ConnectionProxy, userSlugs []string, groupID int64) error {
+	for _, userSlug := range userSlugs {
+		var userID int64
+		err := db.Get(&userID, sq.Select("id").From("users").Where(sq.Eq{"slug": userSlug}))
+		if err != nil {
+			return backend.WrapError("Unable to get user id from slug", backend.BadInputErr(err, fmt.Sprintf(`No user with slug %s was found`, userSlug)))
+		}
+
+		var userGroupMap []models.UserGroupMap
+		err = db.Select(&userGroupMap, sq.Select("*").
+			From("group_user_map").
+			Where(sq.Eq{
+				"user_id":  userID,
+				"group_id": groupID,
+			}))
+		if err != nil {
+			return backend.WrapError("Unable to group from group id", backend.DatabaseErr(err))
+		}
+		if len(userGroupMap) == 0 {
+			_, err = db.Insert("group_user_map", map[string]interface{}{
+				"user_id":  userID,
+				"group_id": groupID,
+			})
+			if err != nil {
+				return backend.WrapError("Unable to connect user to group", backend.DatabaseErr(err))
+			}
+		}
+	}
+
+	return nil
+}
+
+func CreateUserGroup(ctx context.Context, db *database.Connection, i CreateUserGroupInput) (*dtos.UserGroup, error) {
+	if err := isAdmin(ctx); err != nil {
+		return nil, backend.WrapError("Unwilling to create a user group", backend.UnauthorizedReadErr(err))
+	}
+
+	cleanSlug := SanitizeSlug(i.Slug)
+	if cleanSlug == "" {
+		return nil, backend.BadInputErr(errors.New("Unable to create operation. Invalid operation slug"), "Slug must contain english letters or numbers")
+	}
+
+	err := db.WithTx(context.Background(), func(tx *database.Transactable) {
+		id, _ := tx.Insert("user_groups", map[string]interface{}{
+			"slug": cleanSlug,
+			"name": i.Name,
+		})
+		if len(i.UserSlugs) > 0 {
+			AddUsersToGroup(tx, i.UserSlugs, id)
+		}
+	})
+
+	if err != nil {
+		return nil, backend.WrapError("Error creating user group", backend.BadInputErr(err, "A user group with this slug already exists; please choose another name"))
+	}
+	return &dtos.UserGroup{
+		Slug: i.Slug,
+		Name: i.Name,
+	}, nil
+}
```

Comment:
I like this function. it looks super clean.

Reply:
Thanks!